src/core/build_env.go

```diff
@@ -220,38 +184,90 @@ func TestEnvironment(state *BuildState, target *BuildTarget, testDir string) Bui
+		env = append(env, "OUT="+resolveOut(outEnv[0], ".", false))
 	}
 
-	if target.IsTest() {
-		env = append(env,
-			"TOOLS="+strings.Join(toolPaths(state, target.AllTestTools(), true), " "),
-		)
-		if len(target.Test.tools) == 1 {
-			env = append(env, "TOOL="+toolPath(state, target.Test.tools[0], true))
-		}
-		// Named tools as well.
-		for name, tools := range target.Test.namedTools {
-			env = append(env, "TOOLS_"+strings.ToUpper(name)+"="+strings.Join(toolPaths(state, tools, true), " "))
+	return withUserProvidedEnv(target, env)
+}
+
+// ExecEnvironment creates the environment variables for a `plz exec`.
+func ExecEnvironment(state *BuildState, target *BuildTarget, execDir string) BuildEnv {
+	env := append(RuntimeEnvironment(state, target, true, true),
+		"TMP_DIR="+execDir,
+		"TMPDIR="+execDir,
+		"HOME="+execDir,
+		// This is used by programs that use display terminals for correct handling
+		// of input and output in the terminal where the program is run.
+		"TERM="+os.Getenv("TERM"),
+	)
+
+	outEnv := target.Outputs()
+	// OUTS/OUT environment variables being always set is for backwards-compatibility.
+	// Ideally, if the target is a test these variables shouldn't be set.
+	env = append(env, "OUTS="+strings.Join(outEnv, " "))
+	if len(outEnv) == 1 {
+		env = append(env, "OUT="+resolveOut(outEnv[0], ".", target.Sandbox))
+		if target.IsTest() {
+			env = append(env, "TEST="+resolveOut(outEnv[0], ".", target.Test.Sandbox))
 		}
 	}
 
+	return withUserProvidedEnv(target, env)
+}
+
+// RuntimeEnvironment is the base environment for runtime-based environments.
+// Tools and data env variables are made available.
+func RuntimeEnvironment(state *BuildState, target *BuildTarget, abs, inTmpDir bool) BuildEnv {
+	env := TargetEnvironment(state, target)
+
+	if target.Debug != nil {
+		// Debug tools
+		env = append(env, toolsEnv(state, target.AllDebugTools(), target.Debug.namedTools, abs)...)
+	} else if target.IsTest() {
+		// Test tools
+		env = append(env, toolsEnv(state, target.AllTestTools(), target.NamedTestTools(), abs)...)
+	}
+
+	// Data
+	env = append(env, dataEnv(state, target, inTmpDir)...)
+
+	return env
+}
+
+// Handles resolution of OUT files
+func resolveOut(out string, dir string, sandbox bool) string {
+	// Bit of a hack; ideally we would be unaware of the sandbox here.
+	if sandbox && runtime.GOOS == "linux" && !strings.HasPrefix(RepoRoot, "/tmp/") && dir != "." {
+		return path.Join(SandboxDir, out)
+	}
+	return path.Join(dir, out)
+}
+
+// Creates tool-related env variables
+func toolsEnv(state *BuildState, tools []BuildInput, namedTools map[string][]BuildInput, abs bool) BuildEnv {
+	env := BuildEnv{
+		"TOOLS=" + strings.Join(toolPaths(state, tools, abs), " "),
+	}
+	if len(tools) == 1 {
+		env = append(env, "TOOL="+toolPath(state, tools[0], abs))
+	}
+	for name, tools := range namedTools {
+		env = append(env, "TOOLS_"+strings.ToUpper(name)+"="+strings.Join(toolPaths(state, tools, abs), " "))
+	}
+	return env
+}
```

Comment:
This tries to homogenise how tool-related vars are generated for different environments. Both build and test environment would construct both `TOOLS` and `TOOL` differently via a mixture of just relying on unnamed tools or all tools. This approach makes more sense to me and is also aligned with how other vars are currently being handles (i.e. `SRCS`). This might however break build for some users?